internal/nfs/controller/controllerserver.go

```diff
@@ -0,0 +1,141 @@
+	"google.golang.org/grpc/status"
+)
+
+// Server struct of CEPH CSI driver with supported methods of CSI controller
+// server spec.
+type Server struct {
+	csi.UnimplementedControllerServer
+
+	// backendServer handles the CephFS requests
+	backendServer *cephfs.ControllerServer
+}
+
+// NewControllerServer initialize a controller server for ceph CSI driver.
+func NewControllerServer(d *csicommon.CSIDriver) *Server {
+	// global instance of the volume journal, yuck
+	store.VolJournal = journal.NewCSIVolumeJournalWithNamespace(cephfs.CSIInstanceID, fsutil.RadosNamespace)
+
+	return &Server{
+		backendServer: cephfs.NewControllerServer(d),
+	}
+}
+
+// ValidateVolumeCapabilities checks whether the volume capabilities requested
+// are supported.
+func (cs *Server) ValidateVolumeCapabilities(
+	ctx context.Context,
+	req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	return cs.backendServer.ValidateVolumeCapabilities(ctx, req)
+}
+
+// CreateVolume creates the backing subvolume and on any error cleans up any
+// created entities.
+func (cs *Server) CreateVolume(
+	ctx context.Context,
+	req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	res, err := cs.backendServer.CreateVolume(ctx, req)
+	if err != nil {
+		return res, err
+	}
+
+	backend := res.Volume
+
+	log.DebugLog(ctx, "CephFS volume created: %s", backend)
+
+	secret := req.GetSecrets()
+	cr, err := util.NewAdminCredentials(secret)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to retrieve admin credentials: %v", err)
+
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	defer cr.DeleteCredentials()
+
+	nfsVolume, err := NewNFSVolume(res.Volume.VolumeId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	err = nfsVolume.Connect(cr)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to connect: %v", err)
+	}
+
+	err = nfsVolume.CreateExport(backend)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to create export: %v", err)
+	}
+
+	log.DebugLog(ctx, "published NFS-export: %s", nfsVolume)
+
+	// volume has been exported over NFS, set the "share" parameter to
+	// allow mounting
+	backend.VolumeContext["share"] = nfsVolume.GetExportPath()
+
+	return &csi.CreateVolumeResponse{Volume: backend}, nil
+}
+
+// DeleteVolume deletes the volume in backend and its reservation.
+func (cs *Server) DeleteVolume(
+	ctx context.Context,
+	req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	secret := req.GetSecrets()
+	cr, err := util.NewAdminCredentials(secret)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to retrieve admin credentials: %v", err)
+
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	defer cr.DeleteCredentials()
+
+	nfsVolume, err := NewNFSVolume(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	err = nfsVolume.Connect(cr)
```

Comment:
calling Destroy is missing?

Reply:
there was no go-ceph connection yet. With the updated PR go-ceph is used, and `Destroy()` has been added.